### PR TITLE
HWKAPM-637 Cater for null URI in derived events

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
@pavolloffay Could you review please?

Reviewer notes:
* Fix for components not being displayed when URI is not specified (just operation)
* Fix for distributed tracing 'initial endpoints' which were not being displayed due to null uri
* Also found issue with sorted initial endpoints - needed to sort nodes rather than just the info in the dropdown, as caused entry in dropdown to be out of sync with node information in other lists

**Comment by objectiser** [comment]
When viewing the instance list for [query] endpoint, the instance diagrams are not displayed. This has been recorded as a separate jira: https://issues.jboss.org/browse/HWKAPM-643

File: server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java [hunk 1]
```diff
@@ -2212,6 +2231,187 @@ public void testGetCommunicationSummaryStatisticsWithoutOps() {
     }
 
     @Test
+    public void testGetCommunicationSummaryStatisticsWithoutUri() {
+        List<CommunicationDetails> cds = new ArrayList<CommunicationDetails>();
```
Comment (pavolloffay) [inline server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java]: diamond operator?

File: server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java [hunk 2]
```diff
@@ -2212,6 +2231,187 @@ public void testGetCommunicationSummaryStatisticsWithoutOps() {
...
+        List<CompletionTime> cts = new ArrayList<CompletionTime>();
```
Comment (pavolloffay) [inline server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java]: diamond operator?

File: server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java [hunk 3]
```diff
@@ -2212,6 +2231,187 @@ public void testGetCommunicationSummaryStatisticsWithoutOps() {
...
+        CompletionTime ct2_3 = new CompletionTime();
+        ct2_3.setOperation("in2");
+        ct2_3.setBusinessTransaction("testapp");
+        ct2_3.setTimestamp(1700);
+        ct2_3.setDuration(600);
+        cts.add(ct2_3);
+
+        try {
+            analytics.storeFragmentCompletionTimes(null, cts);
+        } catch (Exception e) {
```
Comment (pavolloffay) [inline server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java]: Should be a test error.

File: server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java [hunk 4]
```diff
@@ -2212,6 +2231,187 @@ public void testGetCommunicationSummaryStatisticsWithoutOps() {
...
+        cd5.setBusinessTransaction("testapp");
+        cd5.setTimestamp(1600);
+        cd5.setLatency(500);
+        cd5.setSource(EndpointUtil.encodeEndpoint(null, "in2"));
+        cd5.setTarget(EndpointUtil.encodeEndpoint(null, "out2.1"));
+        cds.add(cd5);
+
+        try {
+            analytics.storeCommunicationDetails(null, cds);
+        } catch (Exception e) {
```
Comment (pavolloffay) [inline server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java]: should be a test error.

File: ui/src/main/scripts/plugins/e2e/ts/e2e.ts [hunk 5]
```diff
@@ -78,12 +79,15 @@ module E2E {
...
       _.each($scope.e2eData, (node) => {
         $scope.topLevel.push(node.id);
-        $scope.topLevel.sort();
         $scope.uris.push(node.uri);
         $scope.operations.push(node.operation);
       });
     };
 
+    $scope.compareNodes = function(a, b) {
+      return a.id > b.id;
```
Comment (pavolloffay) [inline ui/src/main/scripts/plugins/e2e/ts/e2e.ts]: If this goes after https://github.com/hawkular/hawkular-apm/pull/584 it needs to be updated `a.serviceName + ':' + a.id ...`

File: server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java [hunk 6]
```diff
@@ -2212,6 +2231,187 @@ public void testGetCommunicationSummaryStatisticsWithoutOps() {
...
+        assertNotNull(stats);
+        assertEquals(5, stats.size());
+
+        CommunicationSummaryStatistics in1 = null;
+        CommunicationSummaryStatistics in2 = null;
+        CommunicationSummaryStatistics out1_1 = null;
+        CommunicationSummaryStatistics out1_2 = null;
+        CommunicationSummaryStatistics out2_1 = null;
+
+        for (CommunicationSummaryStatistics css : stats) {
```
Comment (pavolloffay) [inline server/elasticsearch/src/test/java/org/hawkular/apm/server/elasticsearch/AnalyticsServiceElasticsearchTest.java]: If there is defined order in `getCommunicationSummaryStatistics` please use a `List` if not maybe is cleaner to use a transformation to a map.

**Comment by objectiser** [comment]
@pavolloffay Updated tests as discussed on IRC, so merging. Thanks!

**Pull request merged**